packages/fluentui/react-northstar/src/utils/positioner/usePopper.ts

```diff
@@ -0,0 +1,391 @@
+      handleStateUpdate,
+      popperOriginalPositionRef,
+    ],
+  );
+}
+
+/**
+ * Exposes Popper positioning API via React hook. Contains few important differences between an official "react-popper"
+ * package:
+ * - style attributes are applied directly on DOM to avoid re-renders
+ * - refs changes and resolution is handled properly without re-renders
+ * - contains a specific to React fix related to initial positioning when containers have components with managed focus
+ *   to avoid focus jumps
+ *
+ * @param {UsePopperOptions} options
+ */
+export function usePopper(
+  options: UsePopperOptions = {},
+): [
+  // React refs are supposed to be contravariant (allows a more general type to be passed rather than a more specific one)
+  // However, Typescript currently can't infer that fact for refs
+  // See https://github.com/microsoft/TypeScript/issues/30748 for more information
+  React.MutableRefObject<any> /* targetRef */,
+  React.MutableRefObject<any> /* containerRef */,
+  React.MutableRefObject<any> /* arrowRef */,
+] {
+  const { enabled = true } = options;
+  const isFirstMount = useFirstMount();
+
+  const popperOriginalPositionRef = React.useRef<string>('absolute');
+  const resolvePopperOptions = usePopperOptions(options, popperOriginalPositionRef);
+
+  const popperInstanceRef = React.useRef<PopperInstance | null>(null);
+
+  const handlePopperUpdate = useEventCallback(() => {
+    popperInstanceRef.current?.destroy();
+    popperInstanceRef.current = null;
+
+    let popperInstance: PopperInstance | null = null;
+
+    if (isBrowser() && enabled) {
+      if (targetRef.current && containerRef.current) {
+        popperInstance = PopperJs.createPopper(
+          targetRef.current,
+          containerRef.current,
+          resolvePopperOptions(targetRef.current, containerRef.current, arrowRef.current),
+        );
+      }
+    }
+
+    if (popperInstance) {
+      /**
+       * The patch updates `.forceUpdate()` Popper function which restores the original position before the first
+       * forceUpdate() call. See also "positionStyleFix" modifier in usePopperOptions().
+       */
+      const originalForceUpdate = popperInstance.forceUpdate;
+
+      popperInstance.isFirstRun = true;
+      popperInstance.forceUpdate = () => {
+        if (popperInstance.isFirstRun) {
+          popperInstance.state.elements.popper.style['position'] = popperOriginalPositionRef.current;
+          popperInstance.isFirstRun = false;
+        }
+
+        originalForceUpdate();
+      };
+    }
+
+    popperInstanceRef.current = popperInstance;
+  });
+
+  // Refs are managed there via useCallbackRef() this allows us to handle properly following scenarios:
+  // - updates for a targetRef without re-rendering
+  // - workarounds refs resolution on React's layer: refs are resolved in the same order as effects that causes an
+  //   issue when you have a container inside a target, for example: actions in ChatMessage
```

Comment:
Do you mind providing more details on this issue ? As someone with less background knowledge I'm  still puzzled what exactly the issue is

Reply:
Agree that it can be non-obvious, thanks for feedback 👍 I completely reworked this section, now it should be much cleaner 🐱

Reply:
Yeah the example with the nested refs helps a lot with the explanation thx